reward/staking_manager.go

```diff
@@ -39,66 +39,102 @@ type blockChain interface {
-		logger.Debug("StakingInfoCache hit.", "Block number", blockNum, "number of staking block", stakingBlockNumber)
+	// Get staking info from cache
+	if cachedStakingInfo := sm.stakingInfoCache.get(stakingBlockNumber); cachedStakingInfo != nil {
+		logger.Debug("StakingInfoCache hit.", "blockNum", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", cachedStakingInfo)
 		return cachedStakingInfo
 	}
 
-	stakingInfo, err := sm.updateStakingCache(stakingBlockNumber)
-	if err != nil {
-		logger.Error("Failed to get stakingInfo", "Block number", blockNum, "number of staking block", stakingBlockNumber, "err", err)
+	// Get staking info from DB
+	if sm.stakingInfoDB != nil {
+		if storedStakingInfo, err := sm.stakingInfoDB.ReadStakingInfo(stakingBlockNumber); storedStakingInfo != nil && err == nil {
+			s, ok := storedStakingInfo.(*StakingInfo)
+			if ok {
+				logger.Debug("StakingInfoDB hit.", "blockNum", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", s)
+				return s
+			} else {
+				logger.Warn("Cannot convert to StakingInfo. Wrong data is written in DB.")
+			}
+		} else {
+			logger.Warn("Failed to get staking info from db.", "blockNum", blockNum, "err", err)
+		}
+	}
+
+	// Get staking info from block header and updates it to cache and db
+	calcStakingInfo, err := sm.updateStakingInfo(stakingBlockNumber)
+	if calcStakingInfo == nil && err != nil {
+		logger.Error("Failed to get stakingInfo", "blockNum", blockNum, "staking block number", stakingBlockNumber, "err", err, "stakingInfo", calcStakingInfo)
 		return nil
 	}
 
-	logger.Debug("Complete StakingInfoCache update.", "Block number", blockNum, "number of staking block", stakingBlockNumber)
-	return stakingInfo
+	logger.Debug("Get stakingInfo from header.", "blockNum", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", calcStakingInfo)
+	return calcStakingInfo
 }
 
 func (sm *StakingManager) IsActivated() bool {
 	return sm.isActivated
 }
 
-// updateStakingCache updates staking cache with a stakingInfo of a given block number.
-func (sm *StakingManager) updateStakingCache(blockNum uint64) (*StakingInfo, error) {
-	stakingInfo, err := sm.ac.getStakingInfoFromAddressBook(blockNum)
+// updateStakingInfo updates staking info in cache and db created from given block number.
+func (sm *StakingManager) updateStakingInfo(blockNum uint64) (*StakingInfo, error) {
+	stakingInfo, err := sm.addressBookConnector.getStakingInfoFromAddressBook(blockNum)
 	if err != nil {
 		return nil, err
 	}
 	sm.isActivated = true
-	sm.sic.add(stakingInfo)
 
-	logger.Info("Add a new stakingInfo to the stakingInfoCache", "stakingInfo", stakingInfo)
+	// update cache
+	sm.stakingInfoCache.add(stakingInfo)
+
+	// update db
+	if sm.stakingInfoDB != nil {
+		err := sm.stakingInfoDB.WriteStakingInfo(blockNum, stakingInfo)
+		if err != nil {
+			logger.Warn("Failed to write staking info to db.", "err", err, "stakingInfo", stakingInfo)
+			return stakingInfo, err
+		}
+	}
```

Comment:
If updating DB can return an error, we should update cache later.

Reply:
What's the reason for your opinion?
In this code, the cache can contain recent information even though DB failure.
Or do you mean If updating DB can return an error, we should update **DB** later?